staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/envelope.go

```diff
@@ -117,22 +170,30 @@ func (t *envelopeTransformer) TransformFromStorage(data []byte, context value.Co
 
 // TransformToStorage encrypts data to be written to disk using envelope encryption.
 func (t *envelopeTransformer) TransformToStorage(data []byte, context value.Context) ([]byte, error) {
-	newKey, err := generateKey(32)
-	if err != nil {
-		return nil, err
-	}
+	var (
+		err         error
+		transformer value.Transformer
+	)
 
-	encKey, err := t.envelopeService.Encrypt(newKey)
-	if err != nil {
-		// Do NOT wrap this err using fmt.Errorf() or similar functions
-		// because this gRPC status error has useful error code when
-		// record the metric.
-		return nil, err
+	// TODO(immutableT) Proactively generate a new DEK when the expiration period is
+	// approaching some threshold (ex. 15%). Fail-open before the expiration period, and
+	// instrument the failure, so that an alert could be generated.
+	if t.currentDEK.shouldRotate() {
+		if err = t.currentDEK.rotate(t.envelopeService, t.dekCacheTTL); err != nil {
+			return nil, err
+		}
+
+		// Cache the currentDEK, so it could be used during decryption.
+		if _, err = t.addTransformer(t.currentDEK.ciphertext, t.currentDEK.plaintext); err != nil {
```

Comment:
`t.currentDEK.cipher/plaintext` should be accessed with a read lock.
add new methods to `dek` to access these instead.